core/src/main/scala/kafka/log/FileMessageSet.scala

```diff
@@ -341,9 +341,11 @@ class FileMessageSet private[kafka](@volatile var file: File,
     if(targetSize > originalSize || targetSize < 0)
       throw new KafkaException("Attempt to truncate log segment to " + targetSize + " bytes failed, " +
                                " size of this log segment is " + originalSize + " bytes.")
-    channel.truncate(targetSize)
-    channel.position(targetSize)
-    _size.set(targetSize)
+    if (targetSize != channel.size().toInt) {
```

Comment:
Let's see what Jun says, but the current code makes sense to me given this. A few things:

1. Should we also throw an exception with an appropriate message `if (targetSize > channel.size)`? And if we do that, it may be clearer to say `targetSize < channel.size` in the current if clause.
2. It's probably worth updating the method documentation to mention whatever changes we have done to the implementation (e.g. we only truncate in some cases).
3. Even though the file has a number of `channel.size().toInt`, it doesn't follow our code convention (it should be `channel.size.toInt`).